### PR TITLE
[Kl-437] As a user i want to see the sorted assets by value in my tokens widget

### PR DESCRIPTION
**Pull request by thesvyatoslavtt**

### CODE REVIEW

File: libs/ui/src/hooks/use-sort-tokens-by-balance.hook.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import { isDefined } from '@rnw-community/shared';
+import { useMemo } from 'react';
+
+import { Token } from '../interfaces/token.interface';
+import { useTokensMarketInfoSelector } from '../store/tokens-market-info/token-market-info.selectors';
+import { useSelectedNetworkSelector } from '../store/wallet/wallet.selectors';
+import { getDollarValue } from '../utils/get-dollar-amount.util';
+import { getTokenMetadataSlug } from '../utils/token-metadata.util';
+
+export const useSortAccountTokensByBalance = (tokens: Token[]): Token[] => {
```
Comment (MuKoJla) [inline libs/ui/src/hooks/use-sort-tokens-by-balance.hook.ts]: you return `Token[]`, but `dollarBalance` is also returned here

File: libs/ui/src/components/navigator/components/component-with-navigation-context/component-with-navigation-context.tsx [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { FC } from 'react';
+
+import { useActiveTokenList } from '../../hooks/use-active-token-list.hook';
+
+export const ComponentWithNavigationContext: FC = () => {
```
Comment (MuKoJla) [inline libs/ui/src/components/navigator/components/component-with-navigation-context/component-with-navigation-context.tsx:6]: just run an app from zero in extension and have this 
image

File: libs/ui/src/screens/tokens/tokens.tsx [hunk 3]
```diff
@@ -108,7 +111,6 @@ export const Tokens: FC = () => {
       return (
         <AccountToken
           token={token}
-          loadBalance={!searchValue}
           showButton={showButton}
           theme={TokenItemThemesEnum.Secondary}
           marketInfo={allTokensMarketInfo[tokenMetadataSlug]}
```
Comment (MuKoJla) [inline libs/ui/src/screens/tokens/tokens.tsx]: if token has `dollarBalance`, so you maybe we do not need to send `marketInfo` here

**Review by MuKoJla — DISMISSED** [review]

**Review by Boltman92 — DISMISSED** [review]

File: libs/ui/src/components/navigator/hooks/use-active-token-list.hook.ts [hunk 4]
```diff
@@ -1,27 +1,48 @@
...
-      dispatch(addNewTokensAction.submit({ debankId, publicKeyHash }));
+      dispatch(getAllUserTokensAction.submit({ debankId, publicKeyHash }));
     }
   };
 
-  useTimerEffect(getActiveTokenList, DATA_UPDATE_TIME, [chainId, publicKeyHash]);
+  useTimerEffect(
+    () => {
+      if (isDefined(currentScreen) && !SCREENS_WITHOUT_UPDATE.includes(currentScreen)) {
+        getActiveTokenList();
```
Comment (unixvb) [inline libs/ui/src/components/navigator/hooks/use-active-token-list.hook.ts]: what is the point of `getActiveTokenList` variable if it's not reused?

File: libs/ui/src/components/navigator/hooks/use-active-token-list.hook.ts [hunk 5]
```diff
@@ -1,27 +1,48 @@
...
+const SCREENS_WITHOUT_UPDATE = [ScreensEnum.Token];
+
 export const useActiveTokenList = () => {
   const dispatch = useDispatch();
   const publicKeyHash = useSelectedAccountPublicKeyHashSelector();
   const { chainId } = useSelectedNetworkSelector();
 
+  const navigationState = useNavigationState(state => state);
+
+  const currentScreen = useMemo(() => {
```
Comment (unixvb) [inline libs/ui/src/components/navigator/hooks/use-active-token-list.hook.ts]: don't we still have hook for `currentScreen`?

**Review by unixvb — CHANGES_REQUESTED** [review]
refactor this PR: 
- it's hard to understand what's going on here;
- a lot of data changes (I think most of them are redundant);
- weird data structure;
- usage of magic strings;
- code not reused.

File: libs/ui/src/components/navigator/hooks/use-active-token-list.hook.ts [hunk 6]
```diff
@@ -3,25 +3,35 @@ import { useDispatch } from 'react-redux';
...
-      dispatch(addNewTokensAction.submit({ debankId, publicKeyHash }));
-    }
-  };
+  useTimerEffect(
+    () => {
+      if (isDefined(currentRoute) && !SCREENS_WITHOUT_UPDATE.includes(currentRoute.name as ScreensEnum)) {
+        const debankId = getDebankId(chainId);
 
-  useTimerEffect(getActiveTokenList, DATA_UPDATE_TIME, [chainId, publicKeyHash]);
+        if (isDefined(debankId) && isNotEmptyString(publicKeyHash)) {
```
Comment (unixvb) [inline libs/ui/src/components/navigator/hooks/use-active-token-list.hook.ts]: combine this conditions

File: libs/ui/src/store/wallet/wallet.epics.ts [hunk 7]
```diff
@@ -88,11 +88,11 @@ const sendAssetEpic = (action$: Observable<Action>, state$: Observable<RootState
 
 const addNewTokensEpic = (action$: Observable<Action>) =>
   action$.pipe(
-    ofType(addNewTokensAction.submit),
+    ofType(getAllUserTokensAction.submit),
     toPayload(),
     concatMap(({ debankId, publicKeyHash }) =>
       from(getTokenList(publicKeyHash, debankId)).pipe(
-        map(tokenList => addNewTokensAction.success({ tokenList, debankGasTokenName: debankId }))
+        map(tokenList => getAllUserTokensAction.success({ tokenList, debankGasTokenName: debankId }))
```
Comment (unixvb) [inline libs/ui/src/store/wallet/wallet.epics.ts:95]: `debankGasTokenName: debankId` wtf?

Comment (thesvyatoslavtt) [reply libs/ui/src/store/wallet/wallet.epics.ts:95]: `debankGasTokenName` and `debankId` have the same names, and better to understand when is name debankGasTokenName in the reducer, but I can add this in reducer `debankId: debankGasTokenName`

File: libs/ui/src/store/wallet/wallet.actions.ts [hunk 8]
```diff
@@ -43,7 +43,7 @@ export const removeNetworkAction = createAction<{ network: NetworkInterface; isN
 );
 
 export const addNewTokenAction = createAction<TokenExtendedMetadata>('wallet/ADD_NEW_TOKEN');
-export const addNewTokensAction = createActions<
+export const getAllUserTokensAction = createActions<
```
Comment (unixvb) [inline libs/ui/src/store/wallet/wallet.actions.ts:46]: change action type

Comment (unixvb) [inline libs/ui/src/store/wallet/wallet.actions.ts:46]: change action type

File: libs/ui/src/interfaces/token.interface.ts [hunk 9]
```diff
@@ -1,6 +1,11 @@
...
+
 import { AccountToken } from './account-token.interface';
 import { TokenMetadata } from './token-metadata.interface';
 
 export type GasTokenMetadata = Pick<TokenMetadata, 'name' | 'decimals' | 'symbol' | 'thumbnailUri'>;
 
 export type Token = AccountToken & TokenMetadata;
+
+export type TokenWithBalance = Token & { valueInDollar: string };
+export type TokenWithBigNumberBalance = Token & { valueInDollar: BigNumber };
```
Comment (unixvb) [inline libs/ui/src/interfaces/token.interface.ts]: use single type everywhere, 'string' seems to be the beast option

File: libs/ui/src/store/wallet/wallet.reducers.ts [hunk 10]
```diff
@@ -152,31 +152,75 @@ export const walletReducers = createReducer<WalletState>(walletInitialState, bui
         }
       };
     })
-    .addCase(addNewTokensAction.success, (state, { payload: { tokenList, debankGasTokenName } }) => {
+    .addCase(getAllUserTokensAction.success, (state, { payload: { tokenList, debankGasTokenName } }) => {
```
Comment (unixvb) [inline libs/ui/src/store/wallet/wallet.reducers.ts:155]: rewrite this reducer using 1 cycle please

File: libs/ui/src/hooks/use-sort-tokens-by-balance.hook.ts [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+import { useMemo } from 'react';
+
+import { GAS_TOKEN_ADDRESS } from '../constants/defaults';
+import { Token, TokenWithBigNumberBalance } from '../interfaces/token.interface';
+import { useTokensMarketInfoSelector } from '../store/tokens-market-info/token-market-info.selectors';
+import { useSelectedNetworkSelector } from '../store/wallet/wallet.selectors';
+import { getDollarValue } from '../utils/get-dollar-amount.util';
+import { getTokenMetadataSlug } from '../utils/token-metadata.util';
+
+export const useSortAccountTokensByBalance = (tokens: Token[]): TokenWithBigNumberBalance[] => {
```
Comment (unixvb) [inline libs/ui/src/hooks/use-sort-tokens-by-balance.hook.ts]: take a look at `useFiatTotalBalance` maybe you could reuse it somehow

File: libs/ui/src/components/token/account-token/account-token.tsx [hunk 12]
```diff
@@ -1,79 +1,49 @@
...
-    }
-  };
-
-  useTimerEffect(getTokenBalance, DATA_UPDATE_TIME, [rpcUrl, publicKeyHash]);
 
   const handleTokenVisibility = () => dispatch(changeTokenVisibilityAction(token));
 
-  const navigateToTokenDetails = () => navigate(ScreensEnum.Token, { token });
+  const navigateToTokenDetails = () =>
+    navigate(ScreensEnum.Token, { token: { ...token, fiatBalance: fiatBalance.toString() } });
```
Comment (unixvb) [inline libs/ui/src/components/token/account-token/account-token.tsx]: no need of converting  `fiatBalance.toString()`, it is string already

File: libs/ui/src/components/token/account-token/account-token.tsx [hunk 13]
```diff
@@ -1,79 +1,49 @@
...
   const dispatch = useDispatch();
   const { navigate } = useNavigation();
-  const { rpcUrl } = useSelectedNetworkSelector();
-  const publicKeyHash = useSelectedAccountPublicKeyHashSelector();
-  const { price } = marketInfo;
-  const { decimals, thumbnailUri, balance, symbol, name, tokenAddress } = token;
+  const { thumbnailUri, balance, symbol, name, tokenAddress, fiatBalance } = token;
   const isGasToken = checkIsGasToken(tokenAddress);
 
+  const fiatBalanceToDispaly = getFiatBalanceToDisplay(new BigNumber(token.balance.data), fiatBalance);
```
Comment (unixvb) [inline libs/ui/src/components/token/account-token/account-token.tsx]: could you make  `fiatBalance` do be formatted like `getFiatBalanceToDisplay` in the place where `fiatBalance` is defined?

File: libs/ui/src/utils/get-dollar-amount.util.ts [hunk 14]
```diff
@@ -13,18 +15,18 @@ interface GetDollarValueArgs {
...
+  amount: string | number;
   decimals: number;
   price: number | undefined;
   errorValue?: string;
   isNeedToFormat?: boolean;
 }): string;
 
 export function getDollarValue(arg: {
-  amount: string;
+  amount: string | number;
```
Comment (unixvb) [inline libs/ui/src/utils/get-dollar-amount.util.ts]: why `number` type added everywhere? could you remove it?

File: libs/ui/src/screens/token/token.tsx [hunk 15]
```diff
@@ -52,14 +56,26 @@ export const Token: FC<Props> = ({ style }) => {
...
+
   const { price, usdPriceChange24h } = useTokenMarketInfoSelector(tokenAddress, chainId);
+  const { tokenBalance, fiatBalance } = useTokenBalance(tokenAddress, tokenId);
 
-  const balanceFromStore = useTokenBalanceSelector(getTokenSlug(tokenAddress, tokenId));
-  const formattedBalance = getFormattedBalance(balanceFromStore ?? balance.data, decimals);
+  const formattedBalance = getFormattedBalance(tokenBalance.toString() ?? balance.data, decimals);
   const isGasToken = checkIsGasToken(tokenAddress);
-  const usdBalance = getDollarValue({ amount: balance?.data ?? 0, price, decimals });
+  const fiatBalanceToDisplay = getFiatBalanceToDisplay(tokenBalance, fiatBalance);
```
Comment (unixvb) [inline libs/ui/src/screens/token/token.tsx]: Same as here https://github.com/madfish-solutions/oko-wallet/pull/255/files#r1087595326  
there will be no need for `getFiatBalanceToDisplay` here and in other places, if you will prepare this data in shared place

File: libs/ui/src/screens/token/token.tsx [hunk 16]
```diff
@@ -52,14 +56,26 @@ export const Token: FC<Props> = ({ style }) => {
     params: { token }
   } = useRoute<RouteProp<ScreensParamList, ScreensEnum.Token>>();
   const { chainId } = useSelectedNetworkSelector();
+  const dispatch = useDispatch();
 
   const { name, symbol, tokenAddress, decimals, tokenId, thumbnailUri, balance } = token;
+
   const { price, usdPriceChange24h } = useTokenMarketInfoSelector(tokenAddress, chainId);
+  const { tokenBalance, fiatBalance } = useTokenBalance(tokenAddress, tokenId);
```
Comment (unixvb) [inline libs/ui/src/screens/token/token.tsx]: why do you need `fiatBalance` from this hook if you are sending `token: TokenWithFiatBalance` as a navigation param and it has `fiatBalance` property?

File: libs/ui/src/hooks/use-sort-tokens-by-balance.hook.ts [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+
+  const sortedAccountTokens = useMemo(
+    () => [
+      ...tokensWithBalance
+        .map(token => ({ ...token, balance: createEntity(getFormattedBalance(token.balance.data, token.decimals)) }))
+        .sort((a, b) => Number(b.fiatBalance) - Number(a.fiatBalance)),
+      ...tokensWithZeroBalance
+        .map(token => ({ ...token, balance: createEntity(getFormattedBalance(token.balance.data, token.decimals)) }))
+        .sort((a, b) => Number(b.balance.data) - Number(a.balance.data))
+    ],
```
Comment (unixvb) [inline libs/ui/src/hooks/use-sort-tokens-by-balance.hook.ts]: you could rewrite all this code (un to the end of the hook) using 1 `.sort()`

File: libs/ui/src/hooks/use-filter-tokens.hook.ts [hunk 18]
```diff
@@ -1,10 +1,9 @@
 import { useMemo, useState } from 'react';
 
 import { EMPTY_STRING } from '../constants/defaults';
-import { Token } from '../interfaces/token.interface';
 import { filterAccountTokensByValue } from '../screens/tokens/utils/filter-account-tokens-by-value';
 
-export const useFilterAccountTokens = (initialAccountTokens: Token[]) => {
+export const useFilterAccountTokens = (initialAccountTokens: any[]) => {
```
Comment (unixvb) [inline libs/ui/src/hooks/use-filter-tokens.hook.ts]: why any?

File: libs/ui/src/store/wallet/wallet.selectors.ts [hunk 19]
```diff
@@ -119,27 +121,62 @@ export const useAccountAssetsSelector = () =>
...
     () => ({
       ...gasTokenMetadata,
       balance: accountsGasTokens[accountGasTokenSlug] ?? 0,
       tokenAddress: GAS_TOKEN_ADDRESS,
-      isVisible: true
+      isVisible: true,
+      fiatBalance
     }),
-    [chainId, accountsGasTokens, accountGasTokenSlug]
+    [chainId, accountsGasTokens, accountGasTokenSlug, allTokensMarketInfo]
```
Comment (unixvb) [inline libs/ui/src/store/wallet/wallet.selectors.ts]: why do you have `allTokensMarketInfo` which is not used inside hook, seems like it should be replaced with `fiatBalance`

File: libs/ui/src/utils/get-dollar-amount.util.ts [hunk 20]
```diff
@@ -1,10 +1,12 @@
 import { isEmptyString, isDefined } from '@rnw-community/shared';
 import { BigNumber } from 'bignumber.js';
 
+import { TOKEN_DOLLAR_VALUE_PLUG } from '../constants/defaults';
+
 import { formatUnits } from './units.utils';
 
 interface GetDollarValueArgs {
-  amount: string;
+  amount: string | number;
```
Comment (unixvb) [inline libs/ui/src/utils/get-dollar-amount.util.ts]: why `string | number` right now? could you have 1 type?

**Review by unixvb — DISMISSED** [review]

**Review by unixvb — APPROVED** [review]

**Review by MuKoJla — APPROVED** [review]

**Review by vladvn1905 — APPROVED** [review]

**Pull request merged**